RearrangeCarGroup8Movements.py

```diff
@@ -0,0 +1,212 @@
+        for move in moves_list:
+            logging.info(move.__str__())
+
+    def print_moves_console(self, moves_list):
+
+        '''Prints a list of moves
+        Arguments:
+            moves: a list of Movement class instances
+        Returns:
+            None
+        '''
+
+        print(self.order_message)
+
+        for move in moves_list:
+            if self.print_console:
+                print(move)
+            else:
+                logging.info(move.__str__())
+        print('\n')
+
+
+class RearrangeCars(object):
+    '''Class to rearrange the cars and print the moves
+       - Uses car_position_dict to optimize searching the position of a given car
+       - Stores additionally the position of empty spot in current car arrangement
+       self.current_car_order.
+       - Moves are either printed to the console or logged
+       - Print moves in O(n), where n is the number of cars.
+       - More precisely at most n + k moves are taken, where k << n is the amount of movements
+       - where car is placed not to the final place
+    '''
+
+    def __init__(self, init_car_order, final_car_order):
+
+        '''Initializes class instance with initial, final order and
+        stores the position of empty parking space in the current car order
+        Memory is O(n) to store 3 order lists, 1 car_position_dict and a set of
+        not arranged cars
+
+        Storing car_position_dict helps to search indices of car in O(1) instead of O(n)
+        A set of not_arranged_cars helps to now when to stop in O(1) as well as to draw
+        in O(1) unarranged car
+
+        Arguments:
+           init_car_order: a valid non-empty list of cars + 1 \
+           empty_place given in their parking place order
+
+           final_car_order: a valid non-empty list of cars + 1 \
+           empty_place according to the desired order
+
+        Returns:
+            None
+        '''
+
+        self.initial_car_order = init_car_order
+        self.final_car_order = final_car_order
+        self.printer = MovementPrinter(init_car_order, final_car_order, True)
+        self._variables_to_init_state()
+
+    def _move_car_to_empty_slot(self, car_to_move_position):
+
+        '''Arranges the car with car_to_move_position index to the empty slot
+        Arguments:
+            car_to_move_position: the position from where the car is moved
+        Returns:
+            None
+        '''
+
+        car_to_move = self.current_car_order[car_to_move_position]
+        self.current_car_order[self.car_position_dict[0]] = car_to_move
+        self.current_car_order[car_to_move_position] = 0
+
+        self.car_position_dict[car_to_move] = self.car_position_dict[0]
+        self.car_position_dict[0] = car_to_move_position
+
+    def _get_not_arranged_car_index(self):
+
+        '''Finds first car index not yet standing on the final destination
+        '''
+
+        car = random.choice(tuple(self.not_arranged_cars))  # O(1)

```

Comment:
I googled how to pick up in 0(1) the element from the set, this was the proposed solution. Since tuple are immutable I suppose we add tuple operation :)